static/compiler-revision-picker.ts

```diff
@@ -0,0 +1,216 @@
+            // TODO: Disable next button
+        }
+
+        const nextRevision = data.results[0];
+
+        // Insert the option into the picker and set it as active.
+        if (Object.prototype.hasOwnProperty.call(this.tomSelect.options, 'nextRevision')) {
+            this.tomSelect.addOptions([
+                {
+                    revision: nextRevision,
+                    offset: nextOffset,
+                },
+            ]);
+        }
+        this.tomSelect.removeItem(item);
+        this.tomSelect.addItem(nextRevision);
+    }
+
+    initialize(compilerId: string) {
+        this.lastCompilerId = compilerId;
+
+        // eslint-disable-next-line @typescript-eslint/no-this-alias
+        const self = this;
+        this.tomSelect = new TomSelect($(this.domNode).find('select')[0], {
+            valueField: 'revision',
+            labelField: 'revision',
+            searchField: ['revision'],
+            placeholder: '🔍 Select revision...',
+            optgroupField: '$groups',
+            lockOptgroupOrder: true,
+            items: compilerId ? [compilerId] : [],
+            dropdownParent: 'body',
+            closeAfterSelect: true,
+            plugins: ['dropdown_input', 'virtual_scroll'],
+            maxOptions: 1000,
+            firstUrl: function (query) {
+                return '0';
+            },
+            load: async function (query, callback) {
+                const offset = parseInt((this as any).getUrl(query));
+                const res = await searchCompilerLibrary({
+                    compilerId: self.lastCompilerId,
+                    query,
+                    offset,
+                });
+                const data = await res.json();
+                const items = data.results.map((revision, i) => ({
+                    revision,
+                    // Track the offset of the revision in the backend result
+                    // set.
+                    offset: offset + i,
+                    // Group is the first dash-delimited substring, which is
+                    // YYYYMMDD for manyclangs.
+                    $groups: [revision.split('-')[0]],
+                }));
+                const groups = _.uniq(items.map(x => x.$groups[0])).map(x => ({
+                    label: self.formatGroupLabel(x),
+                    value: x,
+                }));
+                if (data.offset + data.results.length < data.total) {
+                    (this as any).setNextUrl(query, (data.offset + data.results.length).toString());
+                }
+                callback(items, groups);
+                if ((this as any).items.length === 0 && items.length > 0) {
+                    (this as any).addItem(items[0].revision);
+                }
+            },
+            shouldLoad: function (query) {
+                return true;
+            },
+            onChange: val => {
+                // TODO(jeremy-rifkin) I don't think this can be undefined.
```

Comment:
Is this really a @jeremy-rifkin TODO? I think that's a bit unkind? :) though he might be OK with this being here.

Reply:
I copy-pasted it from compiler-picker.ts. I also couldn't establish with certainty whether this can be undefined :lol:. TODO removed, done.

Reply:
Haha, I hope to return to these tomselect typing issues at some point